src/core/namespace.c

```diff
@@ -3296,6 +3298,280 @@ bool ns_type_supported(NamespaceType type) {
         return access(ns_proc, F_OK) == 0;
 }
 
+/* In target namespace, unmounts an existing overlayfs at mount_path (if one exists), grabs FD from the
+ * underlying directory, and sets up a new overlayfs mount. Coordinates with parent process over pair_fd:
+ * 1. Sends underlying mount directory (e.g., /etc) fd to parent
+ * 2. Creates and sends new overlay fs fd to parent
+ * 3. Waits for parent to configure layers
+ * 4. Performs final mount at mount_path
+ *
+ * This is used by refresh_extensions_in_namespace() to peel back any existing overlays and reapply them.
+ */
+static int unpeel_mount_and_setup_overlay(int pair_fd, const char *mount_path) {
+        _cleanup_close_ int dir_unpeeled_fd = -1, overlay_fs_fd = -1, mount_fd = -1;
+        int r;
+
+        assert(pair_fd >= 0);
+        assert(mount_path);
+
+        /* Create new OverlayFS and send to parent */
+        overlay_fs_fd = fsopen("overlay", FSOPEN_CLOEXEC);
+        if (overlay_fs_fd < 0)
+                return log_debug_errno(errno, "Failed to create overlay fs for %s: %m", mount_path);
+
+        r = send_one_fd(pair_fd, overlay_fs_fd, 0);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to send overlay fs fd to parent: %m");
+
+        /* Opportunistically unmount the directory to peel back any overlays */
+        r = umount2(mount_path, MNT_DETACH|UMOUNT_NOFOLLOW);
```

Comment:
Shouldn't this happen in a loop, in case there are multiple mounts?

Reply:
also won't this umount for the target, live process too? Shouldn't we create a copy of the mount tree and act on that instead?

Reply:
> Shouldn't this happen in a loop, in case there are multiple mounts?

`unpeel_mount_and_setup_overlay` is called in a loop, which should cover that. 

> also won't this umount for the target, live process too? Shouldn't we create a copy of the mount tree and act on that instead?

I already do a `mount(NULL, "/", NULL, MS_SLAVE|MS_REC, NULL)` after entering the namespace. Would that help?

Reply:
If you enter the target mount namespace and call MS_SLAVE | MS_REC then you alter the mount propagation for all mounts in that mount namespace. If the every mount in that mount namespace has been MS_SHARED before then a umount of /whatever would've propagated to at least the mount namespace it was copied from. That will be prevented by MS_SLAVE | MS_REC.

But calling umount() within the target namespace will still wipe the mount for the caller.

What this code does from a cursory reading is:

* create overlayfs context fd and send to some other process (probably in another more privileged set of namespaces)
* umount whatever is currently mount on top of /whatever
* create a detached copy of the now pristine /whatever and send to some other process
* attach a new overlayfs mount

Obviously in between the umount and the attaching of a new overlayfs mount /whatever will reveal whatever was underneath the old overlayfs mount.

But honestly, I'm not that sure what you're asking me.

If the sender and receiver are in the same set of namespaces then obviously umounting /whatever will wipe it for everyone that's in that mount namespace.

Creating a new mount namespace would prevent that obviously. But note, this only works if the new mount namespace is owned by the same user namespace as the mount namespace it's copied from. Otherwise you will get a locked mount tree on /whatever, meaning you can't unmount it.

Reply:
What we want to do is: from pid 1, update the overlayfs on /etc/ of a running service, while it is running. We need to get the original /etc/ if there's already an old overlayfs, that's what the umounts are for. We want to do it async from pid1, so we fork twice.

In the child, we have permission to access images so we mount them. In the grandchild we enter the target namespace to get the refs to the older /etc/. The question is, how do we undo the old overlays in the target mount namespaces _without_ affecting the running process? Do we create another mount namespace after entering the target's? Or can we create an anonymous copy of the target's /etc/ mount using the new mount api and unpeel it from the copy?

Reply:
> also won't this umount for the target, live process too? Shouldn't we create a copy of the mount tree and act on that instead?

Some brief experimentation suggests that this might be true --- it doesn't look like the mount exchange is being done gracefully.

> Or can we create an anonymous copy of the target's /etc/ mount using the new mount api and unpeel it from the copy?

Would this be with a `open_tree(OPEN_TREE_CLONE | AT_RECURSIVE)`? How would unmounting work when you are working with FDs?

Reply:
> In the child, we have permission to access images so we mount them. In the grandchild we enter the target namespace to get the refs to the older /etc/. The question is, how do we undo the old overlays in the target mount namespaces without affecting the running process? Do we create another mount namespace after entering the target's?

If the "running process" (whatever that is) is in the same mount namespace then unsharing the mount namespace is one way of ensuring that umounts don't affect it, yes. Then you can use `open_tree(OPEN_TREE_CLONE)` to create a detached copy of `/etc`. Note that you need to attach the detached mount somewhere before you can add mounts on top of it. It is currently not possible to mount or umount in anonymous mount namespaces.

> Or can we create an anonymous copy of the target's /etc/ mount using the new mount api and unpeel it from the copy?

No, as you cannot currently unmount in anonymous mount namespaces.

Reply:
> Would this be with a `open_tree(OPEN_TREE_CLONE | AT_RECURSIVE)`? How would unmounting work when you are working with FDs?

You cannot unmount in anonymous mount namespaces currently. But in general, it is possible to `umount("/proc/self/fd/<nr>")`.

Reply:
> If the "running process" (whatever that is) is in the same mount namespace then unsharing the mount namespace is one way of ensuring that umounts don't affect it, yes. Then you can use `open_tree(OPEN_TREE_CLONE)` to create a detached copy of `/etc`

Ok, so in other words just before the umount here we need to create a new mount namespace